imxrt-hal/src/lib.rs

```diff
@@ -53,6 +55,36 @@ pub struct Peripherals {
 
 impl Peripherals {
     pub fn take() -> Option<Self> {
+        //TODO determine where best to initialize some peripherals later,
+        // fixes issues
+        cortex_m::interrupt::free(|_| {
+            unsafe {
+                // Disable watchdog powerdown counter
+                modify_reg!(ral::wdog, WDOG1, WMCR, PDE: PDE_0);
+                modify_reg!(ral::wdog, WDOG2, WMCR, PDE: PDE_0);
+
+                // Disable watchdogs
+                modify_reg!(ral::wdog, WDOG1, WCR, WDE: WDE_0);
+                modify_reg!(ral::wdog, WDOG2, WCR, WDE: WDE_0);
+
+                // 0xD928C520U is the update key
+                if (read_reg!(ral::rtwdog, RTWDOG, CS, CMD32EN == CMD32EN_1)) {
+                    write_reg!(ral::rtwdog, RTWDOG, CNT, 0xD928C520);
+                } else {
+                    write_reg!(ral::rtwdog, RTWDOG, CNT, 0xC520);
+                    write_reg!(ral::rtwdog, RTWDOG, CNT, 0xD928);
+                }
+                write_reg!(ral::rtwdog, RTWDOG, TOVAL, 0xFFFF);
+                modify_reg!(
+                    ral::rtwdog,
+                    RTWDOG,
+                    CS,
+                    CMD32EN: CMD32EN_0,
+                    UPDATE: UPDATE_1
+                );
+            }
```

Comment:
Rather than embedding the watchdog disables into `take`, we could move this behavior into the construction of `Watchdog` handles. Until we fully flesh out an API, we could disable the two watchdogs as an effect of `Watchdog` construction. It could resemble

```rust
pub struct Watchdog<M> {
    regs: ral::wdog::WDOG,
}

impl<M> Watchdog<M>
where
  M: module::Module,
{
    pub(crate) fn new(regs: ral::wdog::WDOG) -> Self {
        ral::modify_reg!(ral::wdog, regs, WMCR, PDE: PDE_0);
        ral::modify_reg!(ral::wdog, regs, WCR, WDE: WDE_0);
        Watchdog { regs }
    }
}
```

Then, in the HAL's `Peripherals`:

```rust
pub struct Peripherals {
    // ...
    wdog1: Watchdog<_1>,
    wdog2: Watchdog<_2>,
}

impl Peripherals {
    pub fn take() -> Option<Self> {
        Some(Peripherals {
            wdog1: Watchdog::new(ral::wdog::WDOG1::take()?),
            wdog2: Watchdog::new(ral::wdog::WDOG2::take()?),
            // ... 
        }
    }
}
```

That's all these watchdog handles need to do right now. Comment may apply to a separate `RTWatchdog` handle. I haven't compiled that code, but I don't think we need an `unsafe` block with this approach.

Is there a need for the interrupt-free context when disabling the watchdogs? Given the guarantees we're providing with the HAL, I think users would require `unsafe` code to enable interrupts, or interact with the watchdog registers, before `take()` returns. Do we need to defend against these types of users?

Reply:
I like the idea of moving the wdog1/wdog2/wdog3 disables into their contructors, its wrapped in a critical section because that's what zephyr and mcuxpresso sdks did, notably in those instances the watchdogs are disabled before main() and still disable interrupts, perhaps NXP authors know something I don't, that would be my presumption.

Reply:
For reference here is how zephyr sets up i.mx rt chips, prior to the kernel and then application main
https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/nxp_imx/rt/soc.c#L256